File: app/src/main/java/at/shockbytes/corey/adapter/ExerciseAdapter.java
package at.shockbytes.corey.adapter;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import at.shockbytes.corey.R;
import at.shockbytes.corey.common.core.adapter.BaseAdapter;
import at.shockbytes.corey.common.core.adapter.helper.ItemTouchHelperAdapter;
import at.shockbytes.corey.common.core.workout.model.Exercise;
import butterknife.Bind;

/**
 * @author Martin Macheiner
 *         Date: 02.12.2015.
 */
public class ExerciseAdapter extends BaseAdapter<Exercise>
        implements ItemTouchHelperAdapter {

    private boolean isItemMovable;

    public ExerciseAdapter(Context context, List<Exercise> data) {
        super(context, data);
        isItemMovable = false;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ViewHolder(inflater.inflate(R.layout.item_exercises, parent, false));
    }

    @Override
    public boolean onItemMove(int from, int to) {

        if (from < to) {
            for (int i = from; i < to; i++) {
                Collections.swap(data, i, i + 1);
            }
        } else {
            for (int i = from; i > to; i--) {
                Collections.swap(data, i, i - 1);
            }
        }
        notifyItemMoved(from, to);
        if (onItemMoveListener != null) {
            onItemMoveListener.onItemMove(data.get(to), from, to);
        }
        return true;
    }

    @Override
    public void onItemMoveFinished() {

    }

    @Override
    public void onItemDismiss(int position) {
        Exercise entry = data.remove(position);
        if (onItemMoveListener != null) {
            onItemMoveListener.onItemDismissed(entry, position);
        }
        notifyItemRemoved(position);
    }

    public void setItemsMovable(boolean isItemMovable) {
        this.isItemMovable = isItemMovable;
        notifyDataSetChanged();
    }

    public void filter(String query) {

        List<Exercise> filtered = filterList(data, query);
        setData(filtered);
    }

    private List<Exercise> filterList(List<Exercise> data, String query) {
        final String lowerCaseQuery = query.toLowerCase();

        final List<Exercise> filteredModelList = new ArrayList<>();
        for (Exercise s : data) {
            final String text = s.getName().toLowerCase();
            if (text.contains(lowerCaseQuery)) {
                filteredModelList.add(s);
            }
        }
        return filteredModelList;
    }

    class ViewHolder extends BaseAdapter<Exercise>.ViewHolder {

        @Bind(R.id.item_exercise_txt_name)
        protected TextView txtName;

        @Bind(R.id.item_exercise_imgview_move)
        protected ImageView imgViewMove;

        public ViewHolder(final View itemView) {
            super(itemView);
        }

        @Override
        public void bind(Exercise item){
            content = item;

            txtName.setText(item.getDisplayName(context));
            imgViewMove.setVisibility(isItemMovable ? View.VISIBLE : View.GONE);
        }
    }

}

File: app/src/main/java/at/shockbytes/corey/adapter/ExerciseAdapter.kt
package at.shockbytes.corey.adapter

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import at.shockbytes.corey.R
import at.shockbytes.corey.common.core.workout.model.Exercise
import at.shockbytes.util.adapter.BaseAdapter
import at.shockbytes.util.adapter.ItemTouchHelperAdapter
import kotterknife.bindView
import java.util.*

/**
 * @author Martin Macheiner
 * Date: 02.12.2015.
 */
class ExerciseAdapter(context: Context, data: List<Exercise>)
    : BaseAdapter<Exercise>(context, data.toMutableList()), ItemTouchHelperAdapter {

    private var isItemMovable: Boolean = false

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseAdapter<Exercise>.ViewHolder {
        return ViewHolder(inflater.inflate(R.layout.item_exercises, parent, false))
    }

    override fun onItemMove(from: Int, to: Int): Boolean {

        if (from < to) {
            for (i in from until to) {
                Collections.swap(data, i, i + 1)
            }
        } else {
            for (i in from downTo to + 1) {
                Collections.swap(data, i, i - 1)
            }
        }
        notifyItemMoved(from, to)
        onItemMoveListener?.onItemMove(data[to], from, to)

        return true
    }

    override fun onItemMoveFinished() {
    }

    override fun onItemDismiss(position: Int) {
        val entry = data.removeAt(position)
        onItemMoveListener?.onItemDismissed(entry, position)

        notifyItemRemoved(position)
    }

    fun setItemsMovable(isItemMovable: Boolean) {
        this.isItemMovable = isItemMovable
        notifyDataSetChanged()
    }

    fun filter(query: String) {
        data = filterList(data, query)
    }

    private fun filterList(data: List<Exercise>, query: String): MutableList<Exercise> {
        val lowerCaseQuery = query.toLowerCase()

        val filteredModelList = ArrayList<Exercise>()
        for (s in data) {
            val text = s.name.toLowerCase()
            if (text.contains(lowerCaseQuery)) {
                filteredModelList.add(s)
            }
        }
        return filteredModelList
    }

    internal inner class ViewHolder(itemView: View) : BaseAdapter<Exercise>.ViewHolder(itemView) {

        private val txtName: TextView by bindView(R.id.item_exercise_txt_name)
        private val imgViewMove: ImageView by bindView(R.id.item_exercise_imgview_move)

        override fun bindToView(t: Exercise) {
            content = t

            txtName.text = t.getDisplayName(context)
            imgViewMove.visibility = if (isItemMovable) View.VISIBLE else View.GONE
        }
    }

}